src/Data/Dup/Internal.hs

```diff
@@ -0,0 +1,20 @@
+{-# LANGUAGE GHCForeignImportPrim #-}
+{-# LANGUAGE MagicHash            #-}
+{-# LANGUAGE UnboxedTuples        #-}
+{-# LANGUAGE UnliftedFFITypes     #-}
+
+module Data.Dup.Internal (
+    dup#
+  ) where
+
+import GHC.Prim
+import GHC.Exts
+
+foreign import prim "dupClosure" dupClosure# ::
+  forall s. Any -> State# s -> (# State# s, Any #)
+
+{-# NOINLINE dup# #-}
+dup# :: forall a s. a -> State# s -> (# State# s, a #)
+dup# a s0 =
+    case dupClosure# (unsafeCoerce# a) s0
```

Comment:
(unsafeCoerce# dupClosure#) seems simpler.

Reply:
Would it not be better to keep the scope of `unsafeCoerce(#)` as small as possible?